opencog/atoms/bind/PatternLink.cc

```diff
@@ -191,6 +194,11 @@ PatternLink::PatternLink(const std::set<Handle>& vars,
 			_pat.mandatory.push_back(h);
 		}
 	}
+
+	// If there are any disconnected optionals, consider each of them as an
+	// optional for all mandatory clauses in the pattern
+	_pat.optionals.insert(discon_opts.begin(), discon_opts.end());
```

Comment:
This doesn't make sense. You start by making a copy of `opts`.  Then you remove everything that was copied into `_pat.optionals`.  Now you add everything that's left over.  It would be easier to just say  `_pat.optionals = opts;`  and not bother with the erase.